iree/turbine/kernel/wave/scheduling/aplp/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+use pyo3::prelude::*;
+use pyo3::types::{PyDict, PyList, PyTuple}; // PyAny removed as it's less relevant now
+use rayon::prelude::*;
+use std::sync::Arc;
+
+mod prune; // Contains the prune_envelope function
+
+// Define PyRawEdge as a tuple struct to directly accept Python tuples.
+// The order of fields in the Python tuple must match the order here:
+// (from_node, to_node, delay, iter_diff)
+#[derive(FromPyObject, Debug, Clone)]
+struct PyRawEdge(u32, u32, f32, f32);
+
+// Internal APLP computation function
+fn compute_aplp_internal(
+    node_count: usize,
+    raw_edges: Vec<PyRawEdge>, // Takes Vec of PyRawEdge (tuple structs)
+) -> Vec<Vec<Vec<(f32, f32)>>> {
+    let mut d_current_vec_vec: Vec<Vec<Vec<(f32, f32)>>> =
+        vec![vec![Vec::new(); node_count]; node_count];
+
+    for i in 0..node_count {
+        d_current_vec_vec[i][i] = prune::prune_envelope(vec![(0.0f32, 0.0f32)]);
+    }
+
+    for edge in raw_edges {
+        let u_idx = edge.0 as usize;
+        let v_idx = edge.1 as usize;
+        let delay = edge.2;
+        let iter_diff = edge.3;
+
+        if u_idx < node_count && v_idx < node_count {
+            let new_path_tuple = (delay, iter_diff);
+            let mut current_paths_at_uv = d_current_vec_vec[u_idx][v_idx].clone();
+            current_paths_at_uv.push(new_path_tuple);
+            d_current_vec_vec[u_idx][v_idx] = prune::prune_envelope(current_paths_at_uv);
+        }
+    }
+
+    let mut d_current_arc = Arc::new(d_current_vec_vec);
+
+    for k_idx in 0..node_count {
```

Comment:
For my own educational purposes, is this n^3 computation, essentially iteratively computing all the possible/candidate edges(delay,iter_diff) between all nodes(that has a path between them ofc) and then we prune out that are not in the convex hull since it is not useful for our longest path computation?

Reply:
yes that's correct.